### PR TITLE
Re-organise the modules

### PR DESCRIPTION
**Pull request by davesnx**
As discussed here: https://github.com/reasonml/reason-react/issues/770

This file explains the new organisation of modules: https://github.com/reasonml/reason-react/blob/Unwrap-and-minimize-toplevels/src/README.md

- `wrapped false` and inlining most of the modules makes the interface a little cleaner for the users. On the other side, we have huge modules (which isn't so bad, but that it just my opinion)

- Moved all interface from `React.rei` into it's own module

- I didn't took any decision on `ErrorBoundary`, neither `ReasonReactRouter`. I got the feeling that we should rename those as `SimpleErrorBoundary` and `SimpleRouter` or similar. I would be fine to rename `ReasonReactRouter` to `Router`

### CODE REVIEW

File: src/README.md [hunk 1]
```diff
@@ -1,17 +1,22 @@
-Welcome to the source of ReasonReact!
+### Welcome to the source of ReasonReact
+
+We want to expose as minimum modules to the toplevel as possible, and at the same time want to map closely to the npm packages. So each module maps to a npm package, and each sub-module maps to a sub-module. For example: `react-dom` -> `ReactDOM` and `react-dom/server` -> `ReactDOM.Server`.
 
 Files overview:
 
 ## Bindings
```
Comment (anmonteiro) [inline src/README.md:7]: thanks for writing this up. it's very clear. 

you may want to check the rendering before merging, as the nested bullets aren't rendered as you'd expect.

File: src/README.md [hunk 2]
```diff
@@ -1,17 +1,22 @@
...
+  |- `ReactDOM.Style`: bindings to create `style` objects
+  |- `ReactDOM.TestUtils`: helpers for testing your components
 
-## Extra
+## Extra (not part of React)
 
-- `ReactTestUtils`: helpers for testing your components
-- `ReasonReactErrorBoundary`: component to catch errors within your component tree
 - `ReasonReactRouter`: a simple, yet fully featured router with minimal memory allocations
+- `ErrorBoundary`: component to catch errors within your component tree
```
Comment (anmonteiro) [inline src/README.md:20]: should this one become `React.ErrorBoundary`, or do you want it to stay out at the toplevel?

File: src/React.re [hunk 3]
```diff
@@ -139,10 +456,428 @@ module Suspense = {
...
+[@mel.module "react"]
+external useMemo6:
+  ([@mel.uncurry] (unit => 'any), ('a, 'b, 'c, 'd, 'e, 'f)) => 'any =
+  "useMemo";
+[@mel.module "react"]
+external useMemo7:
+  ([@mel.uncurry] (unit => 'any), ('a, 'b, 'c, 'd, 'e, 'f, 'g)) => 'any =
+  "useMemo";
+
+/* This is used as return values  */
```
Comment (anmonteiro) [inline src/React.re:693]: This was for `type callback('input, 'output)`. which is no longer here

```suggestion
/* This is used as return values  */
```

File: src/React.rei [hunk 4]
```diff
@@ -195,7 +195,7 @@ external useSyncExternalStore:
 [@mel.module "react"]
 external useSyncExternalStoreWithServer:
   (
-    ~subscribe: (([@mel.uncurry] (unit => unit)) => ([@mel.uncurry] (unit => unit))),
+    ~subscribe: ([@mel.uncurry] (unit => unit)) => [@mel.uncurry] (unit => unit),
```
Comment (anmonteiro) [inline src/React.rei:198]: refmt did this?

File: src/dune [hunk 5]
```diff
@@ -1,6 +1,10 @@
 (library
  (name react)
  (public_name reason-react)
+ (wrapped false)
+ ; Explicitly adding modules isn't necessary, but it's a good idea
```
Comment (anmonteiro) [inline src/dune:5]: great idea 👍

**Review by anmonteiro — COMMENTED** [review]
looking very nice

File: src/React.rei (hunk 4 above)
Comment (davesnx) [reply src/React.rei:198]: Yes

File: src/React.re (hunk 3 above)
Comment (davesnx) [reply src/React.re:693]: Removed this and added into the React.Uncurried callback

File: src/React.rei (hunk 4 above)
Comment (davesnx) [reply src/React.rei:198]: Tested and looks fine: https://melange.re/unstable/playground/?language=Reason&code=W0BtZWwubW9kdWxlICJyZWFjdCJdCmV4dGVybmFsIHVzZVN5bmNFeHRlcm5hbFN0b3JlOgogICgKICAgIH5zdWJzY3JpYmU6IChbQG1lbC51bmN1cnJ5XSAodW5pdCA9PiB1bml0KSkgPT4gW0BtZWwudW5jdXJyeV0gKHVuaXQgPT4gdW5pdCksCiAgICB%2BZ2V0U25hcHNob3Q6IHVuaXQgPT4gJ3NuYXBzaG90LAogICAgfmdldFNlcnZlclNuYXBzaG90OiBbQG1lbC51bmN1cnJ5XSAodW5pdCA9PiAnc25hcHNob3QpCiAgKSA9PgogICdzbmFwc2hvdCA9CiAgInVzZVN5bmNFeHRlcm5hbFN0b3JlIjsKCmxldCBnZXRTdGF0ZSA9ICgpID0%2BIDIyOwpsZXQgc3Vic2NyaWJlV2l0aE1vY2sgPSBsaXN0ZW5lciA9PiB7CiAgbGV0IF8gPSBsaXN0ZW5lcigpOwogIEpzLmxvZygiY29zaXRhcyIpOwogICgpID0%2BICgpOwp9OwoKbGV0IHNuYXBzaG90ID0KICB1c2VTeW5jRXh0ZXJuYWxTdG9yZSh%2Bc3Vic2NyaWJlPXN1YnNjcmliZVdpdGhNb2NrLCB%2BZ2V0U25hcHNob3Q9Z2V0U3RhdGUpOwo%3D&live=off

File: src/README.md (hunk 2 above)
Comment (anmonteiro) [reply src/README.md:20]: OK then I'm fine taking the time to think about this, and keeping it `ReasonReactErrorBoundary` for now?

**Review by anmonteiro — APPROVED** [review]
decide what to name `errorBoundary.re` -- I propose `ReasonReactErrorBoundary` to give it a long, unique name.

thanks for working on this

File: src/README.md (hunk 2 above)
Comment (davesnx) [reply src/README.md:20]: Sounds perfect

**Comment by davesnx** [comment]
We would need to make a proper release notes with the module changes, but should be small.

**Comment by joprice** [comment]
@davesnx I'm seeing react-dom/test-utils and react-dom/server getting pulled in by vite during development with this change. Without some sort of change in melange to only import referenced submodules, maybe new modules like `ReactDOMServer` and `ReactDOMTestUtils` are necessary to avoid this?

**Comment by davesnx** [comment]
Splitting those into modules might make sense to avoid this issue. Would this problem fade away in production builds, where those modules would be gone with tree-shaking?

I would like to now how other bundlers handle dev/prod optimisations like this one, to decide if it's a good solution

**Comment by joprice** [comment]
I didn't get that far, since vite didn't like the `require` calls in the server-side commonjs modules. Not sure these particular ones will be shaken away, since I think purity of the module affect tree-shaking as well and would have to check that. I think either way, it's unexpected to pull in test utils or, in general, something you didn't intend to import, even in dev.

**Pull request merged**